@here/harp-mapview/test/TextElementsRendererTest.ts

```diff
@@ -0,0 +1,279 @@
+            mapView.update = resolve;
+        });
+        let numRenderedTextElements = textElementsRenderer.renderTextElements(
+            textElements,
+            time,
+            frameNumber,
+            zoomLevel,
+            renderedTextElements,
+            secondChanceTextElements
+        );
+
+        // First call just triggers the text canvas initialization. Nothing will be rendered.
+        expect(numRenderedTextElements).to.be.equal(0);
+        expect(renderedTextElements.length).to.be.equal(0);
+
+        // TextElementsRenderer.renderTextElements is loading the charset lazily
+        // (FontCatalog.loadCharset) and triggers a mapView.update afterwards.
+        await updatePromise;
+
+        // First "real" frame. Label should be invisble and start fading in.
+        renderedTextElements.splice(0);
+        numRenderedTextElements = textElementsRenderer.renderTextElements(
+            textElements,
+            time + 100,
+            frameNumber + 1,
+            zoomLevel,
+            renderedTextElements,
+            secondChanceTextElements
+        );
+        expect(numRenderedTextElements).to.be.equal(1);
+        expect(renderedTextElements.length).to.be.equal(1);
+        expect(renderedTextElements[0].textRenderState!.opacity).to.be.equal(0);
+        assert(renderedTextElements[0].textRenderState!.isFadingIn());
+
+        // Second frame. Label should be semi-transparent and fading in.
+        renderedTextElements.splice(0);
+        numRenderedTextElements = textElementsRenderer.renderTextElements(
+            textElements,
+            time + 200,
+            frameNumber + 2,
+            zoomLevel,
+            renderedTextElements,
+            secondChanceTextElements
+        );
+
+        expect(numRenderedTextElements).to.be.equal(1);
+        expect(renderedTextElements.length).to.be.equal(1);
+        expect(renderedTextElements[0].textRenderState!.opacity)
+            .to.be.greaterThan(0)
+            .to.be.lessThan(1);
+        assert(renderedTextElements[0].textRenderState!.isFadingIn());
+
+        const lastOpacity = renderedTextElements[0].textRenderState!.opacity;
+
+        // Third frame. Label should be semi-transparent and still fading in.
+        renderedTextElements.splice(0);
+        numRenderedTextElements = textElementsRenderer.renderTextElements(
+            textElements,
+            time + 800,
+            frameNumber + 8,
+            zoomLevel,
+            renderedTextElements,
+            secondChanceTextElements
+        );
+        expect(numRenderedTextElements).to.be.equal(1);
+        expect(renderedTextElements.length).to.be.equal(1);
+        expect(renderedTextElements[0].textRenderState!.opacity)
+            .to.be.greaterThan(lastOpacity)
+            .to.be.lessThan(1);
+        assert(renderedTextElements[0].textRenderState!.isFadingIn());
+
+        // Forth frame. Label should be visible and fading should be finished.
```

Comment:
Why does it take 4 frames to fade in? I tried to find where this is configured but couldn't, is it some default?

Reply:
It is not the number of frames that matter, but the time. It is configurable but by default it is:
`export const DEFAULT_FADE_TIME = 800;` defined in TextElement.ts`
I will use this constant to make things more clear.